### PR TITLE
Pass shortcode atts to action hooks in shortcode templates. Closes #455.

### LINKED ISSUE #455: Pass shortcode attributes as parameter for action hooks in shortcode templates
**Endgoal:** Add shortcode parameters to hooks in shortcode templates. 

We talked about adding the recurring donations automatically to the `[charitable-my-donations]` via the `charitable_my_donations_after` hook. But since there will also be a `[charitable-my-recurring-donations]` shortcode, some users may need/want to keep the recurring donations from displaying via `[charitable-my-donations]` (like with the `hide_login` parameter in the shortcodes that requiring logging in). 

So I wanted to add support for `[charitable-my-donations hide_recurring=1]` from Recurring, which is possible via the `shortcode_atts_charitable_my_donations` filter. 

However, there's currently no way to retrieve that `hide_recurring` attribute from the `charitable_my_donations_after` hook because it isn't passed to the shortcode output and then not passed from there to the output's various hooks. 

What do you think of passing the shortcode `$args|$atts` as part of the `$view_args` to each template so that in the templates I could add the `$atts` to each `add_action` hook?

I was thinking something like this in the my donations shortcode class

```
$view_args = array(
	'donations' => $donations,
	'user'      => $user,
	'atts'		=> $args // shortcode attributes
);
charitable_template( 'shortcodes/my-donations.php', $view_args );
```

and then in the `shortcodes/my-donations.php` template either:
```
/**
 * @hook    charitable_my_donations_after
 * @param object[] $donations An array of donations as a simple object.
 * @param  array $view_args All args passed to template. 
 */
do_action( 'charitable_my_donations_after', $donations, $view_args );
```
Where the shortcode atts are included as a key in the `$view_args` array or parsing out the attributes and just passing them as their own array:

```
$atts = $view_args['atts']`;
/**
 * @hook    charitable_my_donations_after
 * @param object[] $donations An array of donations as a simple object.
 * @param  array $atts All shortcode attributes.
 */
do_action( 'charitable_my_donations_after', $donations, $atts );
```

### PR DESCRIPTION
**Pull request by helgatheviking**
Proposal to include the shortcode atts in the template `$view_args` and then added to all shortcode action hooks. Let me know if you'd prefer the templates' action hooks simply pass the shortcode atts and not all the `$view_args` .

### CODE REVIEW

File: includes/shortcodes/class-charitable-donors-shortcode.php [hunk 1]
```diff
@@ -91,6 +91,9 @@ public static function display( $atts ) {
 					'hide_if_no_donors',
 			) ), $args );
 
+			// Add the shortcode atts to $view_args.
+			$view_args = array_merge( $view_args, array( 'atts' => $args ) );
```
Comment (ericnicolaas) [inline includes/shortcodes/class-charitable-donors-shortcode.php]: Like with the [campaigns] shortcode one, this is not really needed here as we already provide the `charitable_donors_shortcode_view_args` filter on the previous line, so if people need more of the original args passed along, they can use that filter to ensure they get passed along.

File: includes/shortcodes/class-charitable-login-shortcode.php [hunk 2]
```diff
@@ -46,7 +46,10 @@ public static function display( $atts = array() ) {
 
 				ob_start();
 
-				charitable_template( 'shortcodes/logged-in.php', $args );
+				// Add the shortcode atts to $view_args.
+				$view_args = array_merge( $args, array( 'atts' => $args ) );
```
Comment (ericnicolaas) [inline includes/shortcodes/class-charitable-login-shortcode.php]: This seems a bit redundant to me. `$args` is already going to be passed; this just duplicates it in the `atts` property.

File: includes/shortcodes/class-charitable-logout-shortcode.php [hunk 3]
```diff
@@ -42,7 +42,10 @@ public static function display( $atts = array() ) {
 
 			$args = shortcode_atts( $defaults, $atts, 'charitable_logout' );
 
-			charitable_template( 'shortcodes/logout.php', $args );
+			// Add the shortcode atts to $view_args.
+			$view_args = array_merge( $args, array( 'atts' => $args ) );
```
Comment (ericnicolaas) [inline includes/shortcodes/class-charitable-logout-shortcode.php]: As above, this is just copying `$args` back into itself.

File: includes/shortcodes/class-charitable-login-shortcode.php [hunk 4]
```diff
@@ -78,7 +81,10 @@ public static function display( $atts = array() ) {
 
 			ob_start();
 
-			charitable_template( 'shortcodes/login.php', $args );
+			// Add the shortcode atts to $view_args.
+			$view_args = array_merge( $args, array( 'atts' => $view_args ) );
```
Comment (ericnicolaas) [inline includes/shortcodes/class-charitable-login-shortcode.php]: Same issue as above.

File: includes/shortcodes/class-charitable-my-donations-shortcode.php [hunk 5]
```diff
@@ -100,7 +101,7 @@ public static function display( $atts ) {
 			 * @param array  $view_args The view arguments.
 			 * @param array  $args      The query arguments.
 			 */
-			return apply_filters( 'charitable_my_donations_shortcode', ob_get_clean(), $view_args, $args );
+			return apply_filters( 'charitable_my_donations_shortcode', ob_get_clean(), $view_args, $query_args );
```
Comment (ericnicolaas) [inline includes/shortcodes/class-charitable-my-donations-shortcode.php]: 👍 on this one :)

**Comment by helgatheviking** [comment]
@ericnicolaas I *think* the reason was for uniformity between hooks but I can't remember now.

**Comment by ericnicolaas** [comment]
@helgatheviking Yep,  and that makes sense! I'm always keen for improved consistency. :) Just not sure about the redundancies.

**Comment by helgatheviking** [comment]
Looks like you already released version 1.5.5? So these need bumping again right?

**Comment by ericnicolaas** [comment]
@helgatheviking Sorry, yeah. I had to push out a quick release on the weekend due to a pretty major bug (PayPal IPN issue).

**Comment by ericnicolaas** [comment]
Actually, we're up to 1.5.7 for the next release; 1.5.6 is the current.

**Comment by helgatheviking** [comment]
Bumped again.

**Pull request merged**